### PR TITLE
GetComponentIncorrectTypeAnalyzer

### LINKED ISSUE #16: Warn about GetComponent with an incorrect type
#### Problem statement
Sometimes people use GetComponent with types ~~or interfaces~~ which are not Components, which does not work.

#### Proposed solution
No obvious code fix.

### PR DESCRIPTION
**Pull request by AnBucyk**
Fixes #16

#### Checklist

- [X] I have read the [Contribution Guide](/CONTRIBUTING.md) ;
- [X] There is an approved issue describing the change when contributing a new analyzer or suppressor ;
- [x] I have added tests that prove my fix is effective or that my feature works ;
- [x] I have added necessary documentation (if appropriate) ;

#### Short description of what this resolves:
Adds an analyzer that detects if the user is passing a type to `GetComponent` that is not a Unity Component or Interface type

#### TODO
- [X] Detect calls to GetComponent methods in the generic case where the type passed is not a Component or an Interface
- [X] ~~Do we care about the non-generic version of these calls, since we have UNT0003?~~
- [x] Write tests
- [x] Write docs

### CODE REVIEW

File: src/Microsoft.Unity.Analyzers/GetComponentIncorrectType.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			if (symbol.Symbol == null)
+				return;
+
+			if (!IsGetComponent(symbol.Symbol, out var identifier))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, invocation.GetLocation(), identifier));
+		}
+
+		private static bool IsGetComponent(ISymbol symbol, out string identifier)
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers/GetComponentIncorrectType.cs]: This method has too many responsibilities right now and can be safely split in two separate methods:

1) Checking that the invocation expression is indeed a GetComponent call.
2) Checking that the type argument goes against the rule of the diagnostic.

File: src/Microsoft.Unity.Analyzers/GetComponentIncorrectType.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+				return false;
+
+			var containingType = method.ContainingType;
+			if (!containingType.Matches(typeof(UnityEngine.Component)) && !containingType.Matches(typeof(UnityEngine.GameObject)))
+				return false;
+
+			if (!MethodNames.Contains(method.Name))
+				return false;
+
+			var argumentType = method.TypeArguments.First();
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers/GetComponentIncorrectType.cs]: Let's be extra safe and use FirstOrDefault, and bail out if there's none.

**Comment by AnBucyk** [comment]
I'm open to suggestions for more edge cases for tests. Most of the ones I can think of would trigger UNT0003
